geopy/geocoders/geoclient.py

```diff
@@ -0,0 +1,156 @@
+            self,
+            app_id=None,
+            app_key=None,
+            domain='api.cityofnewyork.us/geoclient/v1',
+            timeout=DEFAULT_TIMEOUT,
+            proxies=None,
+            user_agent=None,
+    ):
+        """
+        Initialize Geoclient geocoder. Please note that 'scheme' parameter is
+        not supported: at present state, all Geoclient traffic use https.
+
+        :param string appd_id: The application ID.
+
+        :param string app_key: The application key. 
+
+        :param string domain: Currently it is 'api.cityofnewyork.us/geoclient/v1', can
+            be changed if you have an on-premise Geoclient instance.
+
+        :param dict proxies: If specified, routes this geocoder's requests
+            through the specified proxy. E.g., {"https": "192.0.2.0"}. For
+            more information, see documentation on
+            :class:`urllib2.ProxyHandler`.
+
+        """
+        super(Geoclient, self).__init__(
+            scheme="https", timeout=timeout, proxies=proxies, user_agent=user_agent
+        )
+
+        self.app_id = app_id
+        self.app_key = app_key
+        
+        if not app_id:
+            raise ConfigurationError(
+                'No app_id given, required for api access.  If you do not '
+                'have a Geoclient app_id, sign up here: '
+                'https://developer.cityofnewyork.us/api/geoclient-api'
+                )
+        
+        if not app_key:
+            raise ConfigurationError(
+                'No app_key given, required for api access.  If you do not '
+                'have a Geoclient app_key, sign up here: '
+                'https://developer.cityofnewyork.us/api/geoclient-api'
+                )
+        
+        self.domain = domain.strip('/')
+        
+        '''use SFS (Single-field search) endpoint'''
+        self.geocode_api = 'https://%s/search.json' % (self.domain)
+
+    def geocode(
+            self,
+            query,
+            exactly_one=True,
+            timeout=None,
+            addressdetails=False,
+        ):
+        """
+        Geocode a location query.
+
+        :param string query: The query string to be geocoded using SFS syntax;
+            this must be URL encoded. Documentation at:
+            https://api.cityofnewyork.us/geoclient/v1/doc#section-1.3
+
+        :param bool exactly_one: Return one result or a list of results, if
+            available.
+
+        :param int timeout: Time, in seconds, to wait for the geocoding service
+            to respond before raising a :class:`geopy.exc.GeocoderTimedOut`
+            exception. Set this only if you wish to override, on this call
+            only, the value set during the geocoder's initialization.
+
+        :param bool addressdetails: If you want in *Location.raw* to include
+            address details such as BIN#, community district, etc set it to True.
+            Data Dictionary documenting all available fields at:
```

Comment:
Geoclient returns a LOT of information which can also change with each release (about 4x a year), and explicitly enumerating all the options may actually make it less maintainable.

For example, a request to geocode '85 Fifth Ave, Manhattan, NY', returns:
```JSON
{
  "id": "ip-XXX-XXX-XXX-XXX-XXXXX-XXXXXXXXXX",
  "status": "OK",
  "input": "85 Fifth Ave, Manhattan, NY",
  "results": [
    {
      "level": "0",
      "status": "EXACT_MATCH",
      "request": "address [houseNumber=85, street=Fifth Ave, borough=MANHATTAN, zip=null]",
      "response": {
        "assemblyDistrict": "75",
        "bbl": "1008440001",
        "bblBoroughCode": "1",
        "bblTaxBlock": "00844",
        "bblTaxLot": "0001",
        "bikeLane": "2",
        "boardOfElectionsPreferredLgc": "1",
        "boePreferredStreetName": "5 AVENUE",
        "boePreferredstreetCode": "11041001",
        "boroughCode1In": "1",
        "buildingIdentificationNumber": "1016080",
        "censusBlock2000": "2004",
        "censusBlock2010": "2004",
        "censusTract1990": "  52  ",
        "censusTract2000": "  52  ",
        "censusTract2010": "  52  ",
        "cityCouncilDistrict": "02",
        "civilCourtDistrict": "09",
        "coincidentSegmentCount": "1",
        "communityDistrict": "105",
        "communityDistrictBoroughCode": "1",
        "communityDistrictNumber": "05",
        "communitySchoolDistrict": "02",
        "condominiumBillingBbl": "0000000000",
        "congressionalDistrict": "12",
        "cooperativeIdNumber": "0000",
        "cornerCode": "NE",
        "crossStreetNamesFlagIn": "E",
        "dcpCommercialStudyArea": "11019",
        "dcpPreferredLgc": "01",
        "dotStreetLightContractorArea": "1",
        "dynamicBlock": "203",
        "electionDistrict": "008",
        "fireBattalion": "06",
        "fireCompanyNumber": "003",
        "fireCompanyType": "L",
        "fireDivision": "01",
        "firstBoroughName": "MANHATTAN",
        "firstStreetCode": "11041001020",
        "firstStreetNameNormalized": "FIFTH AVENUE",
        "fromLionNodeId": "0020737",
        "fromPreferredLgcsFirstSetOf5": "0101",
        "genericId": "0001685",
        "geosupportFunctionCode": "1B",
        "geosupportReturnCode": "00",
        "geosupportReturnCode2": "00",
        "gi5DigitStreetCode1": "10410",
        "gi5DigitStreetCode2": "17310",
        "giBoroughCode1": "1",
        "giBoroughCode2": "1",
        "giBuildingIdentificationNumber1": "1016080",
        "giBuildingIdentificationNumber2": "1016080",
        "giDcpPreferredLgc1": "01",
        "giDcpPreferredLgc2": "01",
        "giHighHouseNumber1": "85",
        "giHighHouseNumber2": "3",
        "giLowHouseNumber1": "85",
        "giLowHouseNumber2": "1",
        "giSideOfStreetIndicator1": "R",
        "giSideOfStreetIndicator2": "L",
        "giStreetCode1": "11041001",
        "giStreetCode2": "11731001",
        "giStreetName1": "5 AVENUE",
        "giStreetName2": "EAST   16 STREET",
        "healthArea": "5700",
        "healthCenterDistrict": "15",
        "highBblOfThisBuildingsCondominiumUnits": "1008440001",
        "highCrossStreetB5SC1": "117330",
        "highCrossStreetB5SC2": "134110",
        "highCrossStreetCode1": "11733001",
        "highCrossStreetCode2": "13411001",
        "highCrossStreetName1": "EAST   17 STREET",
        "highCrossStreetName2": "WEST   17 STREET",
        "highHouseNumberOfBlockfaceSortFormat": "000095000AA",
        "houseNumber": "85",
        "houseNumberIn": "85",
        "houseNumberSortFormat": "000085000AA",
        "hurricaneEvacuationZone": "X",
        "instructionalRegion": "MS",
        "interimAssistanceEligibilityIndicator": "I",
        "internalLabelXCoordinate": "0986424",
        "internalLabelYCoordinate": "0207887",
        "latitude": 40.737415391891616,
        "latitudeInternalLabel": 40.73727812604426,
        "legacySegmentId": "0032800",
        "lionBoroughCode": "1",
        "lionBoroughCodeForVanityAddress": "1",
        "lionFaceCode": "2470",
        "lionFaceCodeForVanityAddress": "2470",
        "lionKey": "1247000120",
        "lionKeyForVanityAddress": "1247000120",
        "lionSequenceNumber": "00120",
        "lionSequenceNumberForVanityAddress": "00120",
        "listOf4Lgcs": "01",
        "longitude": -73.9925809709183,
        "longitudeInternalLabel": -73.99215518677124,
        "lowBblOfThisBuildingsCondominiumUnits": "1008440001",
        "lowCrossStreetB5SC1": "117310",
        "lowCrossStreetB5SC2": "134090",
        "lowCrossStreetCode1": "11731001",
        "lowCrossStreetCode2": "13409001",
        "lowCrossStreetName1": "EAST   16 STREET",
        "lowCrossStreetName2": "WEST   16 STREET",
        "lowHouseNumberOfBlockfaceSortFormat": "000085000AA",
        "lowHouseNumberOfDefiningAddressRange": "000085000AA",
        "nta": "MN13",
        "ntaName": "Hudson Yards-Chelsea-Flatiron-Union Square",
        "numberOfCrossStreetB5SCsHighAddressEnd": "2",
        "numberOfCrossStreetB5SCsLowAddressEnd": "2",
        "numberOfCrossStreetsHighAddressEnd": "2",
        "numberOfCrossStreetsLowAddressEnd": "2",
        "numberOfEntriesInListOfGeographicIdentifiers": "0002",
        "numberOfExistingStructuresOnLot": "0001",
        "numberOfStreetFrontagesOfLot": "02",
        "physicalId": "0001916",
        "policePatrolBoroughCommand": "1",
        "policePrecinct": "013",
        "returnCode1a": "00",
        "returnCode1e": "00",
        "roadwayType": "1",
        "rpadBuildingClassificationCode": "O6",
        "rpadSelfCheckCodeForBbl": "9",
        "sanbornBoroughCode": "1",
        "sanbornPageNumber": "034",
        "sanbornVolumeNumber": "02",
        "sanitationCollectionSchedulingSectionAndSubsection": "1B",
        "sanitationDistrict": "105",
        "sanitationRecyclingCollectionSchedule": "ET",
        "sanitationRegularCollectionSchedule": "TTHS",
        "sanitationSnowPriorityCode": "P",
        "segmentAzimuth": "060",
        "segmentIdentifier": "0032800",
        "segmentLengthInFeet": "00245",
        "segmentOrientation": "N",
        "segmentTypeCode": "U",
        "sideOfStreetIndicator": "R",
        "sideOfStreetOfVanityAddress": "R",
        "splitLowHouseNumber": "000085000AA",
        "stateSenatorialDistrict": "28",
        "streetName1In": "FIFTH AVE",
        "streetStatus": "2",
        "taxMapNumberSectionAndVolume": "10306",
        "toLionNodeId": "0020745",
        "toPreferredLgcsFirstSetOf5": "0101",
        "trafficDirection": "A",
        "underlyingStreetCode": "11041001",
        "uspsPreferredCityName": "NEW YORK",
        "workAreaFormatIndicatorIn": "C",
        "xCoordinate": "0986306",
        "xCoordinateHighAddressEnd": "0986405",
        "xCoordinateLowAddressEnd": "0986286",
        "xCoordinateOfCenterofCurvature": "0000000",
        "yCoordinate": "0207937",
        "yCoordinateHighAddressEnd": "0208120",
        "yCoordinateLowAddressEnd": "0207906",
        "yCoordinateOfCenterofCurvature": "0000000",
        "zipCode": "10003"
      }
    }
  ],
  "parseTree": null,
  "policy": null
}
`